test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataReaderExtensionsTests.cs

```diff
@@ -0,0 +1,333 @@
+                {
+                    Name = "Orders",
+                    Target = order,
+                    TargetMultiplicity = EdmMultiplicity.Many
+                });
+
+            EdmEntityContainer defaultContainer = new EdmEntityContainer("NS", "Container");
+            Model.AddElement(defaultContainer);
+            EdmEntitySet customers = defaultContainer.AddEntitySet("Customers", customer);
+            EdmEntitySet orders = defaultContainer.AddEntitySet("Orders", order);
+            customers.AddNavigationTarget(orderNav, orders);
+            customers.AddNavigationTarget(ordresNav, orders);
+        }
+
+        [Fact]
+        public void ReadResourceOrResourceSet_ThrowsArgumentNull_EdmType()
+        {
+            // Arrange & Act & Assert
+            ODataReader reader = null;
+            ExceptionAssert.ThrowsArgumentNull(() => reader.ReadResourceOrResourceSet(), "reader");
+        }
+
+        [Fact]
+        public async Task ReadResourceOrResourceSetAsync_ThrowsArgumentNull_EdmType()
+        {
+            // Arrange & Act & Assert
+            ODataReader reader = null;
+            await ExceptionAssert.ThrowsArgumentNullAsync(() => reader.ReadResourceOrResourceSetAsync(), "reader");
+        }
+
+        [Fact]
+        public async Task ReadResourceWorksAsExpected()
+        {
+            // Arrange
+            const string payload =
+            "{" +
+                "\"@odata.context\":\"http://localhost/$metadata#Customers/$entity\"," +
+                "\"CustomerID\": 7," +
+                "\"Location\": { \"Street\":\"154TH AVE\"}," +
+                "\"Order\": {\"OrderId\": 8, \"Price\": 82 }," +
+                "\"Orders\": []" +
+            "}";
+
+            IEdmEntitySet customers = Model.EntityContainer.FindEntitySet("Customers");
+            Assert.NotNull(customers); // Guard
+
+            // Act
+            Func<ODataMessageReader, Task<ODataReader>> func = mr => mr.CreateODataResourceReaderAsync(customers, customers.EntityType());
+            ODataItemBase item = await ReadPayloadAsync(payload, Model, func);
+
+            // Assert
+            Assert.NotNull(item);
+            ODataResourceWrapper resource = Assert.IsType<ODataResourceWrapper>(item);
+            Assert.Equal(3, resource.NestedResourceInfos.Count);
+            Assert.Equal(new[] { "Location", "Order", "Orders" }, resource.NestedResourceInfos.Select(n => n.NestedResourceInfo.Name));
+        }
+
+        [Fact]
+        public async Task ReadResourceWithPropertyWithoutValueButWithInstanceAnnotationsWorksAsExpected()
+        {
+            // Arrange
+            // Property 'Name' without value but with instance annotations
+            const string payload =
+            "{" +
+                "\"@odata.context\":\"http://localhost/$metadata#Customers/$entity\"," +
+                "\"CustomerID\": 17," +
+                "\"Name@Custom.PrimitiveAnnotation\":123," +
```

Comment:
Could we add a test where the there is a property without value but with instance annotations in a nestedResourceinfo?
For example, in the below case, will the resource.ResourceBase.Properties show the Instance Annotation count as 1? 
or only the NestedResourceInfo 'names' will have the instance annotation count as 1?
{
  "customerId":17,
  "details":{
    "names":{
      "Name@Custom.PrimitiveAnnotation":123
    }
  }
}